### PR TITLE
Encapsulate nested identity loading

### PR DESCRIPTION
**Pull request by romseygeek**
When handling nested documents in the Fetch phase, we currently find
a document's nested identity by loading ObjectMappers from a MapperService,
and iterating over their filters.  This logic is difficult to follow, end requires
exposing internals of the mappers that we would rather hide away.

This commit refactors all NestedIdentity loading into a single NestedDocuments
object, with per-segment iterators to handle the individual documents.  This way
of iterating over documents allows us to share parent filters, meaning that we
can avoid rebuilding the filter for every nested level of every visited document.

### CODE REVIEW

**Comment by romseygeek** [comment]
There are a few other places where we call `MapperService#getObjectMapper` but this I think is the most complex to handle.  It opens the way to some better encapsulation of source loading as well.

One thing to note is that with this change we will add one more bitset to the BitSetFilterCache, for matching child documents - previously we would cache parent and intermediate bitsets, but rebuild a Weight and Scorer for every individual child.  Pulling everything from the BitSetFilterCache means we can avoid regenerating this Weight/Scorer multiple times per fetch, but will end up using a bit more memory; there are probably ways around this if we decide that we want to avoid the extra caching though.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java [hunk 1]
```diff
@@ -706,6 +706,11 @@ public QuerySearchResult queryResult() {
         return queryResult;
     }
 
+    @Override
+    public NestedDocuments getNestedDocuments() {
+        return new NestedDocuments(mapperService(), bitsetFilterCache()::getBitSetProducer);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:711]: looks like if you do indexService.mapperService() you can go ahead and remove mapperService() from SearchContext.

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java]: Done! Good spot.

**Comment by romseygeek** [comment]
> One thing to note is that with this change we will add one more bitset to the BitSetFilterCache, for matching child documents - previously we would cache parent and intermediate bitsets, but rebuild a Weight and Scorer for every individual child. Pulling everything from the BitSetFilterCache means we can avoid regenerating this Weight/Scorer multiple times per fetch, but will end up using a bit more memory; there are probably ways around this if we decide that we want to avoid the extra caching though.

Reading the code more carefully, it seems that we were already caching all object filters (in BitSetFilterCache#BitSetProducerWarmer) but then just ignoring these filters and regenerating them, so this implementation should be a win all round.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch/ci-2

File: server/src/main/java/org/elasticsearch/search/NestedDocuments.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            }
+            if (path == null) {
+                throw new IllegalStateException("Cannot find object path for document " + doc);
+            }
+            SearchHit.NestedIdentity ni = null;
+            int currentLevelDoc = doc;
+            int parentNameLength;
+            while (path != null) {
+                String parent = mapperService.documentMapper().getNestedParent(path);
+                BitSet childBitSet = objectFilters.get(path);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/NestedDocuments.java:154]: We don't really need a bitset here ? The child filter always move forward so we can build the `Scorer` instead and save some memory. That's what the current impl does so I'd prefer that we keep this logic.

**Review by jimczi — CHANGES_REQUESTED** [review]
The refactoring logic looks good to me. I left one comment regarding the change of behavior that needs to be addressed imo.

**Review by jimczi — APPROVED** [review]
LGTM, thanks

**Pull request merged**